File: src/229_Thread_Creation_Patterns/thread_creation.cpp
/*
g++ -pthread --std=c++14 thread_creation.cpp -o app
*/

#include <iostream>
#include <thread>
#include <vector>
#include <string>
#include <chrono>
#include <future>
#include <mutex>

std::mutex mtx;

// Simple worker function that takes parameters
void worker(int id, const std::string& message) {
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    std::lock_guard<std::mutex> lock(mtx);
    std::cout << "Thread " << id << ": " << message << std::endl;
}

// Worker function that returns a value (simulated with shared state)
int calculate_sum(int start, int end) {
    int sum = 0;
    for (int i = start; i <= end; ++i) {
        sum += i;
    }
    return sum;
}

int main() {
    std::cout << "=== Basic Pattern with emplace_back ===" << std::endl;
    
    // Pattern 1: Simple thread creation and joining
    {
        std::vector<std::thread> threads;
        const int num_threads = 5;
        
        // Launch threads using emplace_back
        for (int i = 0; i < num_threads; ++i) {
            threads.emplace_back(worker, i, "Hello from thread");
        }
        
        // Join all threads
        for (auto& t : threads) {
            t.join();
        }
    }
    
    std::cout << "\n=== Pattern with Return Values (using promises) ===" << std::endl;
    
    // Pattern 2: Getting return values using promise/future
    {
        std::vector<std::thread> threads;
        std::vector<std::promise<int>> promises(3);
        std::vector<std::future<int>> futures;
        
        // Create futures from promises
        for (auto& p : promises) {
            futures.push_back(p.get_future());
        }
        
        // Launch threads with lambda to capture promise
        for (int i = 0; i < 3; ++i) {
            threads.emplace_back([](std::promise<int> p, int start, int end) {
                p.set_value(calculate_sum(start, end));
            }, std::move(promises[i]), i * 10, (i + 1) * 10);
        }
        
        // Get results
        for (int i = 0; i < 3; ++i) {
            std::cout << "Sum " << i << ": " << futures[i].get() << std::endl;
        }
        
        // Join threads
        for (auto& t : threads) {
            t.join();
        }
    }
    
    std::cout << "\n=== Pattern with Lambda Capture ===" << std::endl;
    
    // Pattern 3: Using lambdas with capture
    {
        std::vector<std::thread> threads;
        std::vector<int> data = {10, 20, 30, 40, 50};
        
        for (size_t i = 0; i < data.size(); ++i) {
            threads.emplace_back([i, val = data[i]]() {
                std::cout << "Processing index " << i 
                          << " with value " << val * 2 << std::endl;
            });
        }
        
        for (auto& t : threads) {
            t.join();
        }
    }
    
    return 0;
}

File: src/229_Thread_Creation_Patterns/thread_creation.rs
use std::thread;
use std::time::Duration;

// Simple worker function
fn worker(id: usize, message: &str) {
    thread::sleep(Duration::from_millis(100));
    println!("Thread {}: {}", id, message);
}

// Worker function that returns a value
fn calculate_sum(start: i32, end: i32) -> i32 {
    (start..=end).sum()
}

fn main() {
    println!("=== Basic Pattern with JoinHandle ===");
    
    // Pattern 1: Simple thread creation and joining
    {
        let handles: Vec<_> = (0..5)
            .map(|i| {
                thread::spawn(move || {
                    worker(i, "Hello from thread");
                })
            })
            .collect();
        
        // Join all threads
        for handle in handles {
            handle.join().unwrap();
        }
    }
    
    println!("\n=== Pattern with Return Values ===");
    
    // Pattern 2: Getting return values (built into JoinHandle)
    {
        let handles: Vec<_> = (0..3)
            .map(|i| {
                thread::spawn(move || {
                    calculate_sum(i * 10, (i + 1) * 10)
                })
            })
            .collect();
        
        // Collect results
        let results: Vec<i32> = handles
            .into_iter()
            .map(|h| h.join().unwrap())
            .collect();
        
        for (i, sum) in results.iter().enumerate() {
            println!("Sum {}: {}", i, sum);
        }
    }
    
    println!("\n=== Pattern with Move Closure ===");
    
    // Pattern 3: Using move to capture variables
    {
        let data = vec![10, 20, 30, 40, 50];
        
        let handles: Vec<_> = data
            .into_iter()
            .enumerate()
            .map(|(i, val)| {
                thread::spawn(move || {
                    println!("Processing index {} with value {}", i, val * 2);
                })
            })
            .collect();
        
        for handle in handles {
            handle.join().unwrap();
        }
    }
    
    println!("\n=== Scoped Threads (Borrowing Data) ===");
    
    // Pattern 4: Scoped threads for borrowing without moving
    {
        let data = vec![1, 2, 3, 4, 5];
        
        thread::scope(|s| {
            for (i, item) in data.iter().enumerate() {
                s.spawn(move || {
                    println!("Thread {} processing: {}", i, item * 2);
                });
            }
            // All threads automatically joined at scope end
        });
        
        // data is still available here!
        println!("Original data still accessible: {:?}", data);
    }
    
    println!("\n=== Shared State with Mutex ===");
    
    // Pattern 5: Sharing mutable state safely
    {
        use std::sync::{Arc, Mutex};
        
        let counter = Arc::new(Mutex::new(0));
        let handles: Vec<_> = (0..5)
            .map(|_| {
                let counter = Arc::clone(&counter);
                thread::spawn(move || {
                    let mut num = counter.lock().unwrap();
                    *num += 1;
                })
            })
            .collect();
        
        for handle in handles {
            handle.join().unwrap();
        }
        
        println!("Final counter value: {}", *counter.lock().unwrap());
    }
}